src/cmd_sym.ml

```diff
@@ -197,50 +203,58 @@ let cmd profiling debug unsafe optimize workers no_stop_at_failure no_values
   let thread : Thread.t = Thread.create () in
   let results = Choice.run ~workers result thread in
-  let* failing =
-    list_fold_left
-      (fun failing (result, thread) ->
-        let pc = Thread.pc thread in
-        let symbols = thread.symbol_set in
-        let model = get_model ~symbols solver pc in
-        let* result, err =
-          match result with
-          | Symbolic_choice.Multicore.EVal (Ok ()) -> Ok (failing, false)
-          | EAssert assertion ->
-            Ok (`EAssert (assertion, model) :: failing, true)
-          | ETrap tr -> Ok (`ETrap (tr, model) :: failing, true)
-          | EVal (Error e) -> Error e
-        in
-        if not no_values then
-          let testcase =
-            List.sort compare (Encoding.Model.get_bindings model)
-            |> List.map snd
-          in
-          let+ () = write_testcase ~dir:workspace ~err testcase in
-          result
-        else Ok result )
-      [] (List.of_seq results)
-  in
   let print_bug = function
     | `ETrap (tr, model) ->
       Format.pp_std "Trap: %s@\n" (Trap.to_string tr);
-      Format.pp_std "Model:@\n  @[<v>%a@]@\n"
+      Format.pp_std "Model:@\n  @[<v>%a@]@."
         (Encoding.Model.pp ~no_values)
         model
     | `EAssert (assertion, model) ->
       Format.pp_std "Assert failure: %a@\n" Expr.pp assertion;
-      Format.pp_std "Model:@\n  @[<v>%a@]@\n"
+      Format.pp_std "Model:@\n  @[<v>%a@]@."
         (Encoding.Model.pp ~no_values)
         model
   in
-  let rec print_and_count_failures count = function
-    | [] -> count
-    | hd :: tl ->
-      print_bug hd;
-      let count = succ count in
-      if no_stop_at_failure then print_and_count_failures count tl else count
+  let rec print_and_count_failures count_acc results =
+    match results () with
+    | Seq.Nil -> Ok count_acc
+    | Seq.Cons ((result, thread), tl) ->
+      let pc = Thread.pc thread in
+      let symbols = thread.symbol_set in
+      let model = get_model ~symbols solver pc in
+      let* is_err =
+        let open Symbolic_choice.Multicore in
+        match result with
+        | EAssert assertion ->
+          print_bug (`EAssert (assertion, model));
+          Ok true
+        | ETrap tr ->
+          print_bug (`ETrap (tr, model));
+          Ok true
+        | EVal (Ok ()) -> Ok false
+        | EVal (Error e) -> Error e
+      in
+      let count_acc = if is_err then succ count_acc else count_acc in
+      let* () =
+        if not no_values then
+          let testcase =
+            List.sort compare (Encoding.Model.get_bindings model)
+            |> List.map snd
+          in
+          let+ () = write_testcase ~dir:workspace ~err:is_err testcase in
+          ()
+        else Ok ()
+      in
+      if (not is_err) || no_stop_at_failure then
+        print_and_count_failures count_acc tl
+      else Ok count_acc
+  in
+  let results =
+    if deterministic_result_order then
+      List.to_seq @@ List.sort compare @@ List.of_seq results
```

Comment:
I don't understand why we are going back to `Seq` here. The `List.to_seq` could be removed (and this would allow to write `print_and_count_failures` as `list_fold_left` as it was before (adding the count to the accumulator ofc)

Reply:
To maintain laziness if we don't need the deterministic order.

Reply:
I don't understand.

First, here we need the deterministic order for sure because we are in the `then` branch of `if deterministic_result_order then ...`.

Then, running `List.of_seq` already breaks the laziness. And there is no point in going back to `Seq` after that in this case ?

Reply:
Hum OK. Got it. Sorry...